src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs

```diff
@@ -515,6 +591,28 @@ public void AddSourceFile(string fullPath, SourceCodeKind sourceCodeKind = Sourc
             _sourceFiles.AddFile(fullPath, sourceCodeKind, folders);
         }
 
+        public async Task AddSourceFileAsync(string fullPath, SourceCodeKind sourceCodeKind = SourceCodeKind.Regular, ImmutableArray<string> folders = default)
+        {
+            var gate = new SemaphoreSlim(0);
+            void Handler(object sender, string path)
+            {
+                if (path == fullPath)
+                {
+                    gate.Release();
+                }
+            }
```

Comment:
can you place local functions, at the end, after some final return statement?  that way it's clear they're not part of the normal control flow.